tests/unit/Core/Tool/VerifierTest.php

```diff
@@ -0,0 +1,134 @@
+<?php
+
+/**
+ * Unit tests for Signature Verifier
+ *
+ * @author     Ushahidi Team <team@ushahidi.com>
+ * @package    Ushahidi\Application\Tests
+ * @copyright  2013 Ushahidi
+ * @license    https://www.gnu.org/licenses/agpl-3.0.html GNU Affero General Public License Version 3 (AGPL3)
+ */
+
+namespace Tests\Unit\Core\Tool;
+
+use Ushahidi\Core\Tool\Verifier;
+use Ushahidi\Core\Entity\ApiKeyRepository;
+use Tests\TestCase;
+use Mockery as M;
+
+/**
+ * @backupGlobals disabled
+ * @preserveGlobalState disabled
+ */
+class VerifierTest extends TestCase
+{
+	protected $apiKey = "eebc4a87-9267-4491-989c-690fd8f10466";
+	protected $sharedSecret = "iamasharedsecret";
+
+	protected function makeSig($sharedSecret, $url, $payload)
+	{
+		$data = $url . $payload;
+
+		return base64_encode(hash_hmac("sha256", $data, $sharedSecret, true));
+	}
+
+	public function testValidSignature()
+	{
+		$repo = M::mock(ApiKeyRepository::class);
+		$verifier = new Verifier($repo);
+		$url = "http://localhost:8000/api/v3/exports/external/count/1?api_key={$this->apiKey}";
+		$payload = ['data' => 'things'];
+		$signature = $this->makeSig($this->sharedSecret, $url, json_encode($payload));
+
+		$repo->shouldReceive('apiKeyExists')->with($this->apiKey)->andReturn(true);
+
+		$return = $verifier->verified($signature, $this->apiKey, $this->sharedSecret, $url, json_encode($payload));
+
+		$this->assertTrue($return);
+	}
+
+	public function testValidSignatureWithEmptyPayload()
+	{
+		$repo = M::mock(ApiKeyRepository::class);
+		$verifier = new Verifier($repo);
+		$url = "http://localhost:8000/api/v3/exports/external/count/1?api_key={$this->apiKey}";
+		$payload = "";
+		$signature = $this->makeSig($this->sharedSecret, $url, "");
```

Comment:
My brain is dead now, but I think this actually fails if it's not a json_encoded empty array. 🤷‍♂️

Reply:
So I changed the Signer to stop json_encoding everything... because thats weird. We now validate the raw contents of the request body as a string